cranelift/codegen/src/machinst/lower.rs

```diff
@@ -111,29 +109,25 @@ pub trait LowerCtx {
     /// Get the value of a constant instruction (`iconst`, etc.) as a 64-bit
     /// value, if possible.
     fn get_constant(&self, ir_inst: Inst) -> Option<u64>;
-    /// Get the input in any combination of three forms:
+    /// Get the input as one of two options other than a direct register:
     ///
-    /// - An instruction, if the same color as this instruction or if the
-    ///   producing instruction has no side effects (thus in both cases
-    ///   mergeable);
-    /// - A constant, if the value is a constant;
-    /// - A register.
+    /// - An instruction, given that it is effect-free or able to sink its
+    ///   effect to the current instruction being lowered, and given it has only
+    ///   one output, and if effect-ful, given that this is the only use;
+    /// - A constant, if the value is a constant.
     ///
-    /// The instruction input may be available in some or all of these
-    /// forms. More than one is possible: e.g., it may be produced by an
-    /// instruction in the same block, but may also have been forced into a
-    /// register already by an earlier op. It will *always* be available
-    /// in a register, at least.
+    /// The instruction input may be available in either of these forms.  It may
+    /// be available in neither form, if the conditions are not met; if so, use
+    /// `put_input_in_reg()` instead to get it in a register.
     ///
-    /// If the backend uses the register, rather than one of the other
-    /// forms (constant or merging of the producing op), it must call
-    /// `use_input_reg()` to ensure the producing inst is actually lowered
-    /// as well. Failing to do so may result in the instruction that generates
-    /// this value never being generated, thus resulting in incorrect execution.
-    /// For correctness, backends should thus wrap `get_input()` and
-    /// `use_input_regs()` with helpers that return a register only after
-    /// ensuring it is marked as used.
-    fn get_input(&self, ir_inst: Inst, idx: usize) -> LowerInput;
+    /// If the backend merges the effect of a side-effecting instruction, it
+    /// must call `sink_inst()`. When this is called, it indicates that the
+    /// effect has been sunk to the current scan location. The sunk
+    /// instruction's result(s) must have *no* uses remaining, because it will
+    /// not be codegen'd (it has been integrated into the current instruction).
+    fn maybe_get_input_as_source_or_const(&self, ir_inst: Inst, idx: usize) -> LowerInput;
+    /// Get the `idx`th input in a register.
```

Comment:
doc nitpick for clarity:

```suggestion
    /// Put the `idx`th input into a register and return the assigned register.
```

Reply:
Done.